### PR TITLE
ppai: add landcover sample

### PR DESCRIPTION
**Pull request by davidcavazos**
## Description

Add Land Cover sample for Dynamic World.

TODO:
- [x] Make sure tests pass in the testing infrastructure
- [x] Add README.md and update top level README.md
- [x] Add PyDocs and comments
- [x] Document `argparse` command line arguments
- [x] Add notebook text
- [x] Make sure all requirements are at their latest

Staged notebook: https://colab.research.google.com/github/davidcavazos/python-docs-samples/blob/ppai-landcover/people-and-planet-ai/land-cover-classification/README.ipynb

⚠️ To be able to install the requirements since it's not merged yet, run:
```py
github_path = "https://raw.githubusercontent.com/davidcavazos/python-docs-samples/ppai-landcover/people-and-planet-ai/land-cover-classification"
!pip install -r {github_path}/requirements.txt -c {github_path}/constraints.txt
```

⚠️ To be able to clone the repo since it's not merged yet, run:
```py
!git clone -b ppai-landcover https://github.com/davidcavazos/python-docs-samples/
%cd python-docs-samples/people-and-planet-ai/land-cover-classification/
```

## Checklist
- [x] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [x] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [x] **Tests** pass:   `nox -s py-3.6` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [x] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [x] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [x] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [x] Please **merge** this PR for me once it is approved.
- [x] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample

### CODE REVIEW

**Comment by nikitamaia** [comment]
This is an awesome sample! Loved seeing all the different land cover use cases at the end of the notebook.

Can you update the keras.io links in the notebook to be the equivalent ones on tensorflow.org instead?

ie: 
https://keras.io/api/layers/core_layers/input/ ---> https://www.tensorflow.org/api_docs/python/tf/keras/Input

**Comment by davidcavazos** [comment]
I'm glad you liked it. When I get to write the notebook text, I'll make sure to bring in some background of what each of them represents.

**Comment by ivanmkc** [comment]
I'll wait for the README before I review so I have a sense of how the files work together.

**Comment by davidcavazos** [comment]
@ivanmkc The missing README (markdown) only contains a small summary. The actual instructions README is the notebook itself. I'm working on the PyDocs of the source files right now, but at least the instructions and how to use every file should already be there.

File: people-and-planet-ai/land-cover-classification/data/prediction-locations.csv [hunk 1]
```diff
@@ -0,0 +1,37 @@
+name,lat,lon,year
```
Comment (ivanmkc) [inline people-and-planet-ai/land-cover-classification/data/prediction-locations.csv:1]: Shouldn't data should go in gs://cloud-samples-data/vertex-ai/?

File: people-and-planet-ai/land-cover-classification/data/training-bounds.csv [hunk 2]
```diff
@@ -0,0 +1,81 @@
+west,south,east,north
```
Comment (ivanmkc) [inline people-and-planet-ai/land-cover-classification/data/training-bounds.csv:1]: Ditto

File: people-and-planet-ai/land-cover-classification/e2e_test.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+
+    subprocess.check_call(["gsutil", "-m", "rm", "-rf", f"gs://{bucket_name}/*"])
+    bucket.delete(force=True)
+
+
+@pytest.fixture(scope="session")
+def container_image() -> str:
+    # https://cloud.google.com/sdk/gcloud/reference/builds/submit
+    container_image = f"gcr.io/{PROJECT}/{NAME}:{UUID}"
+    subprocess.check_call(
```
Comment (ivanmkc) [inline people-and-planet-ai/land-cover-classification/e2e_test.py:86]: There's also a Python SDK if you want to use that instead.

Comment (ivanmkc) [reply people-and-planet-ai/land-cover-classification/e2e_test.py:86]: I can see how this would be simpler though.

File: people-and-planet-ai/land-cover-classification/serving/main.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+DEFAULT_MODEL_PATH = os.environ["MODEL_PATH"]
+
+
+@app.route("/predict/<float(signed=True):lat>/<float(signed=True):lon>/<int:year>")
+def predict(lat: float, lon: float, year: int) -> flask.Response:
+    """Gets a prediction from the model.
+
+    Args:
+        lat: Latitede of the point of interest.
```
Comment (ivanmkc) [inline people-and-planet-ai/land-cover-classification/serving/main.py:57]: Spelling: Latitude

File: people-and-planet-ai/land-cover-classification/serving/main.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    lon: float,
+    bands: List[str],
+    patch_size: int,
+    scale: int,
+) -> np.ndarray:
+    """Fetches a patch of pixels from Earth Engine as a NumPy array.
+
+    Args:
+        image: Earth Engine image to fetch from.
+        lat: Latitde of the point of interest.
```
Comment (ivanmkc) [inline people-and-planet-ai/land-cover-classification/serving/main.py:177]: Spelling: Latitude

**Comment by davidcavazos** [comment]
Latest failure is a timeout because I pushed two commits too close to each other, but the tests technically passed.

File: people-and-planet-ai/land-cover-classification/e2e_test.py (hunk 3 above)
Comment (davidcavazos) [reply people-and-planet-ai/land-cover-classification/e2e_test.py:86]: I would prefer to leave it like this for simplicity, it also matches more closely what is done in the notebook.

File: people-and-planet-ai/land-cover-classification/serving/main.py (hunk 4 above)
Comment (davidcavazos) [reply people-and-planet-ai/land-cover-classification/serving/main.py]: Fixed

File: people-and-planet-ai/land-cover-classification/serving/main.py (hunk 5 above)
Comment (davidcavazos) [reply people-and-planet-ai/land-cover-classification/serving/main.py]: Fixed

File: people-and-planet-ai/land-cover-classification/data/prediction-locations.csv (hunk 1 above)
Comment (davidcavazos) [reply people-and-planet-ai/land-cover-classification/data/prediction-locations.csv:1]: We've stored other CSV files in GitHub for other samples before. They're small enough to not have a performance impact, plus it gives more visibility for people on what's on the file or if they want to edit them. It also makes it simpler to run locally.

I prefer it like this, but I can upload them to Cloud Storage if you have a strong opinion about it.

File: people-and-planet-ai/land-cover-classification/trainer.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    """
+
+    # Read the training and validation datasets.
+    training_dataset = read_dataset(training_data, patch_size, batch_size)
+    validation_dataset = read_dataset(validation_data, patch_size, batch_size)
+
+    # Create, train and save the model.
+    model = new_model(training_dataset)
+    model.fit(
+        training_dataset.shuffle(10),
```
Comment (nikitamaia) [inline people-and-planet-ai/land-cover-classification/trainer.py:181]: I think you would want to shuffle before you batch. Otherwise you are just shuffling within the batches themselves.

File: people-and-planet-ai/land-cover-classification/serving/main.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+            return f.getvalue()
+
+    except Exception as e:
+        # If something didn't go well, log the error with the stack trace
+        # and return a more descriptive error message to the client.
+        logging.error(e, exc_info=True)
+        return (f"{type(e).__name__}: {e}", 500)
+
+
+def ee_init() -> None:
```
[the remark this answers is not in the document]
Comment (davidcavazos) [reply people-and-planet-ai/land-cover-classification/serving/main.py:115]: Yes, the caveat is that we have to register the service account in Earth Engine, but we had to do that anyways to run in Dataflow. But yes, I can't wait until the authentication flow is similar to other Cloud products.

File: people-and-planet-ai/land-cover-classification/trainer.py (hunk 6 above)
Comment (davidcavazos) [reply people-and-planet-ai/land-cover-classification/trainer.py:181]: If we shuffle before caching, wouldn't that make all the batches the same anyways? The data points should already be in a random order due to Dataflow and the interleaving, right?

Comment (nikitamaia) [reply people-and-planet-ai/land-cover-classification/trainer.py:181]: never mind, shuffling after the batching doesn't shuffle within the individual batches. It just shuffles the order of all the batches. Either way, what is the purpose of doing that, since as you mention the data is already in a random order?

Comment (ivanmkc) [reply people-and-planet-ai/land-cover-classification/trainer.py:181]: I've looked into this a few years ago, here was a good resource for shuffling and randomness: https://www.moderndescartes.com/essays/shuffle_viz/

**Comment by ivanmkc** [comment]
Pixel-wise accuracy (as currently used) doesn't make sense for image segmentation use cases, partially due to highly imbalanced classes. IOU would be a standard metric.

File: people-and-planet-ai/land-cover-classification/trainer.py (hunk 6 above)
Comment (davidcavazos) [reply people-and-planet-ai/land-cover-classification/trainer.py:181]: I guess the idea was to feed the batches in a "random" order so each epoch would be different, but it might not make any difference.

Should I just remove the shuffle? Or do you think it still adds value? Or should we shuffle before batching?

**Comment by davidcavazos** [comment]
<p>Since we're using one-hot encoded labels, should we be using this one?</p><p> </p><p>https://www.tensorflow.org/api_docs/python/tf/keras/metrics/OneHotIoU</p>

---

 <a href='https://app.reviewnb.com/GoogleCloudPlatform/python-docs-samples/pull/7805/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: people-and-planet-ai/land-cover-classification/trainer.py (hunk 6 above)
Comment (davidcavazos) [reply people-and-planet-ai/land-cover-classification/trainer.py]: I removed the shuffle for now, but please let me know if you think we should still do it in any other way.

**Comment by davidcavazos** [comment]
I changed the accuracy metric to the one-hot encoded IoU. I'm not sure if that'll change how the model trains or just the metrics logging, but I'm still retraining the model and will update the prediction images.

**Review by nikitamaia — APPROVED** [review]

**Review by ivanmkc — APPROVED** [review]
LGTM

File: people-and-planet-ai/land-cover-classification/data/prediction-locations.csv (hunk 1 above)
Comment (leahecole) [reply people-and-planet-ai/land-cover-classification/data/prediction-locations.csv:1]: I'm pro keeping it in github if it has no negative performance impact because it's easier maintenance wise

**Review by leahecole — APPROVED** [review]
Thank you for including such detailed comments

File: people-and-planet-ai/land-cover-classification/trainer.py (hunk 6 above)
Comment (ivanmkc) [reply people-and-planet-ai/land-cover-classification/trainer.py]: I think we should visualize the data shuffling to see the optimal shuffling technique. I'll run some simulations.
